File: guests/go/feed/feed.go
// The feed scene from Go: sum-typed elements, end to end. The sealed
// marker interface is the sum, the structs its constructors, and
// elimination is Go-shaped on both sides — typed case arms author the
// template blueprints, and handlers type-switch on the model's current
// value, a refinement the witnessed UpdateField checks rather than
// trusts, so a stale occurrence folds into nothing.
package feed

import (
	"fmt"

	kaya "dev.kaya/bindings/go"
)

// Post is the sum. kaya-gen reads this declaration — the implementing
// structs, IN DECLARATION ORDER — and emits post_kaya.go.
//
//go:generate go run dev.kaya/cmd/kaya-gen -type Post -key string
type Post interface{ isPost() }

type Note struct{ Text string }

type Todo struct {
	Title string
	Done  bool
}

func (Note) isPost() {}
func (Todo) isPost() {}

func App() *kaya.App {
	app := kaya.NewApp()

	app.Build(func(tx *kaya.Tx) {
		feed := PostCollection(tx)
		doneCount := feed.Derive(tx, func(items []kaya.RecordEntry[string, Post]) string {
			n := 0
			for _, e := range items {
				if todo, ok := e.Value.(Todo); ok && todo.Done {
					n++
				}
			}
			return fmt.Sprintf("%d done", n)
		})

		tx.Mount(tx.Row(func() {
			tx.Button("promote", func(tx *kaya.Tx) {
				for _, e := range feed.Items(tx) {
					if note, ok := e.Value.(Note); ok {
						feed.Update(tx, e.Key, Todo{Title: note.Text, Done: true})
						break
					}
				}
			})
			tx.Label(doneCount)
			// The generated eliminator: one required arm per
			// constructor, so a missing arm is a missing argument.
			PostEachSum(tx, feed,
				func(note kaya.SumCase[string, Note]) {
					note.Label(func(n *Note) *string { return &n.Text })
				},
				func(todo kaya.SumCase[string, Todo]) {
					todo.Row(func() {
						todo.Checkbox(func(td *Todo) *bool { return &td.Done },
							func(tx *kaya.Tx, key string, checked bool) {
								// The generated refined patch: the
								// comma-ok re-eliminates at write
								// time, so a stale occurrence folds
								// into the !ok arm.
								if todo, ok := PostAsTodo(tx, feed, key); ok {
									todo.Done(checked)
								}
							})
						todo.Label(func(td *Todo) *string { return &td.Title })
					})
				},
			)
		}))
		for _, ins := range []struct {
			key  string
			post Post
		}{
			{"a", Note{Text: "jot one"}},
			{"b", Todo{Title: "buy milk"}},
			{"c", Note{Text: "jot two"}},
		} {
			feed.Insert(tx, ins.key, ins.post)
		}
	})

	return app
}

File: guests/java/dev/kaya/guests/Feed.java
package dev.kaya.guests;

import dev.kaya.KayaGen;
import dev.kaya.KayaApp;
import dev.kaya.KayaRecords;

/**
 * The feed scene from the JVM — guests/rust/feed.rs, tools/scenes/feed.steps.
 */
public final class Feed {
    /** The annotation processor reads this and generates PostKaya. */
    @KayaGen(key = "String")
    sealed interface Post permits Note, Todo {}

    record Note(String text) implements Post {}

    record Todo(String title, boolean done) implements Post {}

    public static void app() {
        KayaApp app = new KayaApp();

        app.build(tx -> {
            var feed = PostKaya.collection(tx);
            KayaApp.Signal<String> doneCount = feed.derive(tx, items -> {
                int n = 0;
                for (KayaRecords.Entry<String, Post> entry : items) {
                    if (entry.value instanceof Todo todo && todo.done()) {
                        n++;
                    }
                }
                return n + " done";
            });

            tx.mount(tx.row(() -> {
                tx.button("promote", t -> {
                    for (KayaRecords.Entry<String, Post> entry : feed.items(t)) {
                        if (entry.value instanceof Note note) {
                            feed.update(t, entry.key, new Todo(note.text(), true));
                            break;
                        }
                    }
                });
                tx.label(doneCount);
                PostKaya.eachSum(tx, feed)
                            .note((t, note) -> {
                                note.label(t, Note::text);
                            })
                            .todo((t, todo) -> {
                                t.row(() -> {
                                    todo.checkbox(t, Todo::done,
                                            (KayaApp.Tx t2, String key, boolean checked) -> {
                                                // The refined patch re-eliminates at
                                                // write time: a stale occurrence
                                                // folds into the empty.
                                                PostKaya.asTodo(t2, feed, key)
                                                        .ifPresent(p -> p.done(checked));
                                            });
                                    todo.label(t, Todo::title);
                                });
                            });
            }));

            feed.insert(tx, "a", new Note("jot one"));
            feed.insert(tx, "b", new Todo("buy milk", false));
            feed.insert(tx, "c", new Note("jot two"));
            return null;
        });

        app.dispatchLoop();
    }

    private Feed() {}
}
